math/pythagoras.cpp

```diff
@@ -0,0 +1,94 @@
+
+/**
+ * Copyright 2020 @author beqakd
+ * @file
+ *
+ * A basic implementation of 3d distance between two point.
+ */
+
+#include <cassert>   // For testing(assert)
+#include <cmath>     // For std::pow(), std::sqrt() and so on
+#include <iostream>  // For io operations
+
+/**
+ * The distance between two points in a three dimensional - 3D - coordinate
+ * system. For this we have 3 point x, y, z.
+ * Also we created method int distance which takes points and returns distance.
+ */
+template <typename T> class point {
+ private:
+    T x;
+    T y;
+    T z;
+
+ public:
+    /**
+    * We have two constructor. First is for 3D point. it takes T type 3 value
+    * and save it to our calss. Second one takes only 2 T type value and third
+    * one is defoult 0 and save it into the class too.
+    */
+    point(T a, T b, T c) : x(a), y(b), z(c) {}  // Consructor for 3D points
+    point(T a, T b) : x(a), y(b), z(0) {}  // Second constructor for 2D points
+    double distance(const point &);  // Our prototype of distance method
```

Comment:
shouldn't this line be:
```cpp
template<typename T2>
double distance(const point<T2> &other) {
```
notice the `<T2>` for the input argument otherwise, its type could be undefined or implicit typecast?
It is however, compiling using GCC, MSVC and CLANG. wonder what I missed.

Reply:
Its not if you declare method outside of the class like this
template <typename T> double point<T>::distance(const point &other)
So its not problem.